x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportFinalizeJobExecutionAction.java

```diff
@@ -58,20 +75,60 @@ protected void masterOperation(FinalizeJobExecutionAction.Request request, Clust
                                    ActionListener<AcknowledgedResponse> listener) {
 
         MlMetadata mlMetadata = MlMetadata.getMlMetadata(state);
-        List<String> jobsInClusterState = Arrays.stream(request.getJobIds())
+        Set<String> jobsInClusterState = Arrays.stream(request.getJobIds())
                 .filter(id -> mlMetadata.getJobs().containsKey(id))
-                .collect(Collectors.toList());
-
-        // This action should not be called for jobs that have
-        // their configuration in index documents
+                .collect(Collectors.toSet());
 
         if (jobsInClusterState.isEmpty()) {
-            // This action is a no-op for jobs not defined in the cluster state.
-            listener.onResponse(new AcknowledgedResponse(true));
-            return;
+            finalizeIndexJobs(Arrays.asList(request.getJobIds()), listener);
+        } else {
+            ActionListener<AcknowledgedResponse> finalizeClusterStateJobsListener = ActionListener.wrap(
+                    ack -> finalizeClusterStateJobs(jobsInClusterState, listener),
+                    listener::onFailure
+            );
+
+            Set<String> jobsInIndex = new HashSet<>(Arrays.asList(request.getJobIds()));
+            jobsInIndex.removeAll(jobsInClusterState);
+
+            finalizeIndexJobs(jobsInIndex, finalizeClusterStateJobsListener);
+        }
+    }
+
+    private void finalizeIndexJobs(Collection<String> jobIds, ActionListener<AcknowledgedResponse> listener) {
+        String jobIdString = String.join(",", jobIds);
+        logger.debug("finalizing jobs [{}]", jobIdString);
```

Comment:
nit: isn't it stringified to "[id1, id2, ...]" anyway? + less garbage for non-debug?

Reply:
Fair point. I'll remedy in a later commit as this is merged now.